File: Source/Camping Mod 2.0/common/rikmuld/camping/entity/tileentity/TileEntityCampfireCook.java
package rikmuld.camping.entity.tileentity;

import java.util.ArrayList;
import java.util.Arrays;
import java.util.Random;

import net.minecraft.item.ItemStack;
import net.minecraft.nbt.NBTTagCompound;
import net.minecraft.util.AxisAlignedBB;
import rikmuld.camping.core.register.ModItems;
import rikmuld.camping.core.util.ItemStackUtil;
import rikmuld.camping.core.util.PacketUtil;
import rikmuld.camping.inventory.slot.SlotCooking;
import rikmuld.camping.item.ItemParts;
import rikmuld.camping.misc.cooking.CookingEquipment;
import rikmuld.camping.misc.cooking.CookingEquipmentList;
import rikmuld.camping.network.packets.PacketItems;
import cpw.mods.fml.relauncher.Side;
import cpw.mods.fml.relauncher.SideOnly;

public class TileEntityCampfireCook extends TileEntityInventory {

	public int maxFeul = 12000;
	public int fuelForCoal = 600;
	public int fuel;
	public float[][] coals = new float[3][20];
	Random rand = new Random();

	public int[] cookProgress = new int[10];

	public CookingEquipment equipment;
	public ArrayList<SlotCooking> slots;

	private boolean active;
	private boolean oldActive;

	private int update;

	public TileEntityCampfireCook()
	{
		for(int i = 0; i < 20; i++)
		{
			coals[0][i] = rand.nextFloat() / 5F;
			coals[1][i] = rand.nextFloat() / 5F;
			coals[2][i] = rand.nextFloat() * 360;
		}
	}

	private void cookFood()
	{
		if(equipment != null)
		{
			for(int i = 0; i < equipment.maxFood; i++)
			{
				if(fuel > 0)
				{
					if((getStackInSlot(i + 2) != null) && ((getStackInSlot(i + 2).itemID != ModItems.parts.itemID) || (getStackInSlot(i + 2).getItemDamage() != ItemParts.ASH)))
					{
						cookProgress[i]++;
					}
				}
				else
				{
					if(cookProgress[i] > 0)
					{
						cookProgress[i] = 0;
					}
				}
				if((getStackInSlot(i + 2) == null) && (cookProgress[i] > 0))
				{
					cookProgress[i] = 0;
				}

				if(cookProgress[i] >= equipment.cookTime)
				{
					if(equipment.canCook(getStackInSlot(i + 2).itemID, getStackInSlot(i + 2).getItemDamage()))
					{
						setInventorySlotContents(i + 2, equipment.cookableFoood.get(Arrays.asList(getStackInSlot(i + 2).itemID, getStackInSlot(i + 2).getItemDamage())).copy());
					}
					else
					{
						setInventorySlotContents(i + 2, new ItemStack(ModItems.parts, 1, ItemParts.ASH));
					}
					PacketUtil.sendToAllPlayers(new PacketItems(i + 2, xCoord, yCoord, zCoord, getStackInSlot(i + 2)));
					cookProgress[i] = 0;
				}
			}
		}
	}

	public int getCoalPieces()
	{
		return fuel > 0? (((fuel / fuelForCoal) + 1) <= 20? (fuel / fuelForCoal) + 1:20):0;
	}

	@Override
	@SideOnly(Side.CLIENT)
	public AxisAlignedBB getRenderBoundingBox()
	{
		return AxisAlignedBB.getAABBPool().getAABB(xCoord, yCoord, zCoord, xCoord + 1, yCoord + 1, zCoord + 1);
	}

	public float getScaledCoal(int maxPixels)
	{
		return ((float)fuel / (float)maxFeul) * maxPixels;
	}

	public float getScaledcookProgress(int maxPixels, int foodNum)
	{
		return (((float)cookProgress[foodNum] + 1) / equipment.cookTime) * maxPixels;
	}

	@Override
	public int getSizeInventory()
	{
		return 12;
	}

	public void manageCookingEquipment()
	{
		if((equipment == null) && (getStackInSlot(1) != null))
		{
			equipment = CookingEquipmentList.getCooking(getStackInSlot(1));
		}
		else if((equipment != null) && (getStackInSlot(1) == null))
		{
			equipment = null;
		}

		if(slots != null)
		{
			if(equipment != null)
			{
				for(int i = 0; i < equipment.maxFood; i++)
				{
					if(!slots.get(i).active)
					{
						slots.get(i).activate(equipment.slots[0][i], equipment.slots[1][i], equipment.cookableFoood.keySet());
					}
				}
			}

			if(equipment == null)
			{
				for(int i = 0; i < 10; i++)
				{
					if(slots.get(i).active)
					{
						slots.get(i).deActivate();
						if(slots.get(i).getStack() != null)
						{
							ItemStackUtil.dropItemsInWorld(new ItemStack[]{slots.get(i).getStack()}, worldObj, xCoord, yCoord, zCoord);
							slots.get(i).putStack(null);
						}
					}
				}
			}
		}
	}

	public void manageFuel()
	{
		if(fuel > 0)
		{
			fuel--;
			sendTileData(0, true, fuel);
		}
		if(((fuel + fuelForCoal) <= maxFeul) && (getStackInSlot(0) != null))
		{
			decrStackSize(0, 1);
			fuel += fuelForCoal;
		}
	}

	public void manageLight()
	{
		if(active != oldActive)
		{
			update = 3;
		}
		if(update > 0)
		{
			update--;
			worldObj.markBlockForUpdate(xCoord, yCoord, zCoord);
			worldObj.markBlockForRenderUpdate(xCoord, yCoord, zCoord);
		}
	}

	@Override
	public void readFromNBT(NBTTagCompound tag)
	{
		super.readFromNBT(tag);
		fuel = tag.getInteger("fuel");
		cookProgress = tag.getIntArray("cookProgress");

		for(int i = 0; i < coals.length; i++)
		{
			for(int j = 0; j < coals[i].length; j++)
			{
				coals[i][j] = tag.getFloat("coals" + i + j);
			}
		}
	}

	public void setSlots(ArrayList<SlotCooking> slots)
	{
		this.slots = slots;
	}

	@Override
	public void setTileData(int id, int[] data)
	{
		if(id == 0)
		{
			fuel = data[0];
		}
	}

	@Override
	public void updateEntity()
	{
		manageCookingEquipment();
		if(!worldObj.isRemote)
		{
			active = fuel > 0;
			manageFuel();
			cookFood();
			manageLight();
			oldActive = active;
		}
	}

	@Override
	public void writeToNBT(NBTTagCompound tag)
	{
		super.writeToNBT(tag);
		tag.setInteger("fuel", fuel);
		tag.setIntArray("cookProgress", cookProgress);

		for(int i = 0; i < coals.length; i++)
		{
			for(int j = 0; j < coals[i].length; j++)
			{
				tag.setFloat("coals" + i + j, coals[i][j]);
			}
		}
	}
}

File: scala/com/rikmuld/camping/features/blocks/campfire/cook/TileEntityCampfireCook.scala
package com.rikmuld.camping.features.blocks.campfire.cook

import com.rikmuld.camping.CampingMod
import com.rikmuld.camping.CampingMod._
import com.rikmuld.camping.Definitions.CampfireCook._
import com.rikmuld.camping.Definitions.Parts
import com.rikmuld.camping.features.blocks.campfire.Roaster
import com.rikmuld.camping.features.blocks.campfire.cook.equipment.CookingEquipment
import com.rikmuld.corerm.network.PacketSender
import com.rikmuld.corerm.network.packets.PacketItemData
import com.rikmuld.corerm.tileentity.TileEntityInventory
import net.minecraft.block.state.IBlockState
import net.minecraft.entity.player.EntityPlayer
import net.minecraft.item.ItemStack
import net.minecraft.nbt.NBTTagCompound
import net.minecraft.util.ITickable
import net.minecraft.util.math.{AxisAlignedBB, BlockPos}
import net.minecraft.world.World
import net.minecraftforge.fml.relauncher.{Side, SideOnly}

import scala.collection.JavaConversions._
import scala.collection.mutable.ArrayBuffer

object TileEntityCampfireCook {
  def isAsh(stack: ItemStack): Boolean =
    stack.getItem == CampingMod.OBJ.parts && stack.getItemDamage == Parts.ASH
}

class TileEntityCampfireCook extends TileEntityInventory with Roaster with ITickable {

  final val MAX_FUEL =
    CONFIG.campfireMaxFuel

  final val COAL_FUEL =
    CONFIG.campfireCoalFuel

  private var fuel: Int =
    0

  private var cookProgress: Array[Int] =
    new Array[Int](10)

  private var equipment: Option[CookingEquipment] =
    None

  private val guiPlayers: ArrayBuffer[EntityPlayer] =
    new ArrayBuffer()

  @SideOnly(Side.CLIENT)
  override def getRenderBoundingBox: AxisAlignedBB =
    new AxisAlignedBB(pos.getX, pos.getY, pos.getZ, pos.getX + 1, pos.getY + 1, pos.getZ + 1)

  override def shouldRefresh(world: World, pos: BlockPos, old: IBlockState, nw: IBlockState): Boolean =
    old.getBlock != nw.getBlock

  override def readFromNBT(tag: NBTTagCompound): Unit = {
    super.readFromNBT(tag)

    fuel = tag.getInteger("fuel")
    cookProgress = tag.getIntArray("cookProgress")
    equipment = CookingEquipment.getEquipment(getStackInSlot(1))
  }

  override def writeToNBT(tag: NBTTagCompound): NBTTagCompound = {
    tag.setInteger("fuel", fuel)
    tag.setIntArray("cookProgress", cookProgress)

    super.writeToNBT(tag)
  }

  override def canRoast(item: ItemStack): Boolean =
    fuel > 0 && super.canRoast(item)

  override def roastSpeed(item: ItemStack): Float =
    0.75f + (getCoalPieces / 12f)

  override def getSizeInventory: Int =
    12

  override def getName: String =
    "campfire_cooking"

  override def closeInventory(player: EntityPlayer): Unit =
    guiPlayers -= player

  def getCoalPieces(fuel: Int): Int =
    Math.ceil(fuel / COAL_FUEL.toFloat).toInt

  def getCoalPieces: Int =
    getCoalPieces(fuel)

  def getEquipment: Option[CookingEquipment] =
    equipment

  def getScaledFuel(maxPixels: Int): Float =
    (fuel.toFloat / MAX_FUEL) * maxPixels

  def getScaledCookProgress(maxPixels: Int, slot: Int): Float =
    equipment.fold(0f)(equip =>
      (cookProgress(slot).toFloat + 1) / equip.getCookTime
    ) * maxPixels

  def getCookProcess(slot: Int): Int =
    cookProgress(slot)

  def getFuel: Int =
    fuel

  @SideOnly(Side.CLIENT)
  def setFuel(fuel: Int): Unit =
    this.fuel = fuel

  @SideOnly(Side.CLIENT)
  def setCookProcess(slot: Int, data: Int): Unit =
    cookProgress(slot) = data

  override def setTileData(id: Int, data: Seq[Int]): Unit = id match {
    case 0 =>
      fuel = data.head
    case 1 =>
      for(i <- data.indices)
        cookProgress(i) = data(i)
  }

  private def cookFood(): Unit = {
    equipment.foreach(eq => {
      for (i <- 0 until eq.getMaxCookingSlot)
        cookFood(i, getStackInSlot(i + 2), eq)

      guiPlayers.foreach(player =>
        sendTileDataTo(1, player, cookProgress.take(eq.getMaxCookingSlot): _*)
      )
    })
  }

  private def cookFood(i: Int, food: ItemStack, equipment: CookingEquipment): Unit =
    if (fuel == 0 || food.isEmpty || TileEntityCampfireCook.isAsh(food)) cookProgress(i) = 0
    else {
      if (cookProgress(i) >= equipment.getCookTime)
        doCookFood(i, food, equipment)
      else
        cookProgress(i) += 1
    }

  private def doCookFood(i: Int, food: ItemStack, equipment: CookingEquipment): Unit = {
    val result = equipment.getResult(food).getOrElse(
      new ItemStack(CampingMod.OBJ.parts, 1, Parts.ASH)
    )

    cookProgress(i) = 0

    setInventorySlotContents(i + 2, result)
    PacketSender.sendToClient(new PacketItemData(i + 2, pos.getX, pos.getY, pos.getZ, result.copy()))
  }

  override def onChange(slot: Int): Unit =
    if (slot == 1) {
      (equipment, CookingEquipment.getEquipment(getStackInSlot(1))) match {
        case (None, Some(nw)) =>
          equipment = Some(nw)
        case (Some(old), None) =>
          equipment = None
        case (Some(old), Some(nw)) if old != nw =>
          equipment = Some(nw)
        case _ =>
      }
    }

  override def openInventory(player: EntityPlayer): Unit =
    guiPlayers.add(player)

  def addFuel(): Unit = {
    if (fuel == 0)
      getBlock.setState(world, pos, STATE_ON, true)

    decrStackSize(0, 1)

    fuel += COAL_FUEL
  }

  def removeFuel(): Unit = {
    fuel -= 1

    if (fuel == 0 && !hasCoal)
      getBlock.setState(world, pos, STATE_ON, false)
  }

  def fuelTick(): Unit = {
    val oldFuel = fuel

    if (fuel > 0)
      removeFuel()

    if (hasCoal && fuel + COAL_FUEL <= MAX_FUEL)
      addFuel()

    if (fuel != oldFuel)
      fuelChanged(oldFuel, fuel)
  }

  def fuelChanged(old: Int, nw: Int): Unit =
    if (getCoalPieces != getCoalPieces(old))
      sendTileData(0, true, fuel)
    else guiPlayers.foreach(player =>
      sendTileDataTo(0, player, fuel)
    )

  def hasCoal: Boolean =
    !getStackInSlot(0).isEmpty

  def isOn: Boolean =
    fuel > 0

  override def update(): Unit =
    if (!world.isRemote) {
      fuelTick()
      cookFood()
    }
}